src/uthash.h

```diff
@@ -1172,10 +1176,16 @@ typedef struct UT_hash_bucket {
 
 typedef struct UT_hash_table {
    UT_hash_bucket *buckets;
+#ifdef DUTHASH_V2_ABI
+   struct UT_hash_handle *tail; /* tail hh in app order, for fast append    */
+   ptrdiff_t hho; /* hash handle offset (byte pos of hash handle in element */
+#endif
    unsigned num_buckets, log2_num_buckets;
    unsigned num_items;
+#ifndef DUTHASH_V2_ABI
    struct UT_hash_handle *tail; /* tail hh in app order, for fast append    */
    ptrdiff_t hho; /* hash handle offset (byte pos of hash handle in element */
+#endif
```

Comment:
As long as you're changing the struct layout, I wonder whether we should add explicit "padding" fields at the end of the struct, since (with the Bloom filter) it currently contains 8x+5 bytes in the 32-bit case and 8x+1 bytes in the 64-bit case. But I worry that the same people complaining about the difference between 56 bytes and 64 bytes might just as well start complaining about not being able to store things in the tail-padding of this struct. 😛

Reply:
What do you thing about to declare two entire separated structs? For example:

```c
#if UTHASH_ABI_VERSION >= 3
typedef struct UT_hash_table {
  all >=3 members ...
}
#else
typedef struct UT_hash_table {
  all <=2 members ...
}
#endif
```

Using `-Werror=padded/-Wpadded` we can't declare `tail/hho` as tail in the struct because error "`padding struct to align ‘tail’`".

Reply:
By "tail-padding", I meant "padding after the last member of the struct whatever it is," not "padding before/after the member whose name is `tail`."

Reply:
I'm not sure if I understood "padding after the last member of the struct whatever it is" correct, but it compiles:

```bash
gcc main.c
```
`main.c`
```c
struct foo {
    int a;
    char *b;
};

int main() {
    struct foo bar;
    printf("%zd - %zd\n", sizeof(bar.a), sizeof(bar.b));
    return 0;
}
```
it raises error at compile time (because `b` is larger than `a`):

```bash
gcc -Werror=padded main.c 
```
```
main.c:5:11: error: padding struct to align ‘b’ [-Werror=padded]
     char *b;
           ^
cc1: some warnings being treated as errors
```

Reply:
Additional note: the GCC/Clang has the attribute `__attribute__((packed))`, but I don't know how to use it yet (I'm newbie on GCC attributes).

Reply:
The GCC/Clang has the attribute `__attribute__((packed))`, but I'm not attribute expert so I don't know any performance impact by adding such option in the uthash struct. Anyway, the following code compiles fine even using `-Wpadded`:

```c
typedef struct
#if UTHASH_ABI_VERSION >= 3
  __attribute__((packed))
#endif
UT_hash_table {
   UT_hash_bucket *buckets;
   unsigned num_buckets, log2_num_buckets, num_items;
   struct UT_hash_handle *tail;
   ptrdiff_t hho;
   unsigned ideal_chain_maxlen;
...
} UT_hash_table;
```